### PR TITLE
Add proto-lens-discrimination package for Groups and Sorts of message types.

### PR DESCRIPTION
**Pull request by awpr**
Note this currently suffers from https://github.com/ekmett/discrimination/issues/3, so it probably shouldn't be released just yet.

### CODE REVIEW

File: proto-lens-tests/proto-lens-tests.cabal [hunk 1]
```diff
@@ -47,6 +47,26 @@ Test-Suite canonical_test
...
+  build-depends: base
+               , containers
+               , data-default
+               , discrimination
+               , lens-family
+               , HUnit
+               , proto-lens
+               , proto-lens-protoc
+               , proto-lens-arbitrary
+               , proto-lens-discrimination
```
Comment (judah) [inline proto-lens-tests/proto-lens-tests.cabal]: Can we move the tests to a test-suite in the proto-lens-discrimination package, similar to what we do for proto-lens-combinators?  (As in that package, it doesn't really share code (or need to) with the rest of proto-lens-tests.)

Also, along those lines: please add proto-lens-discrimination and discrimination-ieee754 to stack.yaml and travis-cabal.sh so they get tested by travis.

File: discrimination-ieee754/discrimination-ieee754.cabal [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+copyright:           Google Inc.
+category:            Data
+build-type:          Simple
+cabal-version:       >=1.10
+
+library
+  exposed-modules:     Data.Discrimination.IEEE754
+  -- other-modules:       
+  -- other-extensions:    
+  build-depends:       base >=4.8 && <4.9, data-binary-ieee754 >=0.4 && <0.5, discrimination >=0.2 && <0.3, contravariant >=1.4 && <1.5
```
Comment (judah) [inline discrimination-ieee754/discrimination-ieee754.cabal]: Line length; format like the build-depends of the test suite?

File: discrimination-ieee754/test/test.hs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+sortCompare :: Sort a -> a -> a -> Ordering
+sortCompare s l r = case runSort s [(l, LT), (r, GT)] of
+    [x]:_ -> x
+    _     -> EQ
+
+propAgreesWithOrd :: Ord a => Sort a -> (a, a) -> Bool
+propAgreesWithOrd s (l, r) = compare l r == sortCompare s l r
+
+main = defaultMain
```
Comment (judah) [inline discrimination-ieee754/test/test.hs]: Might be worth adding some explicit checks for the corner cases, e.g. 0 and -0, since that specific case is unlikely to be tested by QuickCheck (at least, in a quick test I ran in ghci).

File: discrimination-ieee754/src/Data/Discrimination/IEEE754.hs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+module Data.Discrimination.IEEE754 where
+
+import Data.Binary.IEEE754 (floatToWord, doubleToWord)
+import Data.Bits (xor, setBit, shift)
+import Data.Discrimination (Grouping(..), Group, Sorting(..), Sort)
+import Data.Functor.Contravariant (Contravariant(contramap))
+import Data.Word (Word32, Word64)
+
+-- TODO(awpr): Do these agree with the IEEE754 "total ordering"?
```
Comment (judah) [inline discrimination-ieee754/src/Data/Discrimination/IEEE754.hs]: Maybe clarify if the open question is about NaN behavior or something else?

File: proto-lens-discrimination/proto-lens-discrimination.cabal [hunk 5]
```diff
@@ -0,0 +1,34 @@
+-- Initial proto-lens-discrimination.cabal generated by cabal init.  For
```
Comment (judah) [inline proto-lens-discrimination/proto-lens-discrimination.cabal]: Remove this `cabal init` header

File: proto-lens-discrimination/proto-lens-discrimination.cabal [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+build-type:          Simple
+cabal-version:       >=1.10
+
+library
+  exposed-modules:     Data.ProtoLens.Sort
+                     , Data.ProtoLens.Group
+                     , Data.ProtoLens.Discrimination
+  build-depends:       base >=4.8 && <4.9
+                     , data-default >=0.5 && <0.6
+                     , bytestring >=0.10.6
```
Comment (judah) [inline proto-lens-discrimination/proto-lens-discrimination.cabal]: bytestring==0.10.* (similar to proto-lens itself)
It looks like you're not using anything fancy in proto-lens itself; also we're giving upper bounds to all the other packages.

File: proto-lens-discrimination/proto-lens-discrimination.cabal [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+                     , contravariant >=1.3 && <1.5
+                     , containers >=0.5 && <0.6
+                     , discrimination >=0.2 && <0.3
+                     , discrimination-ieee754 ==0.1.*
+                     , lens-family ==1.2.*
+                     , proto-lens ==0.1.*
+                     , text >=1.2
```
Comment (judah) [inline proto-lens-discrimination/proto-lens-discrimination.cabal]: maybe: text == 1.2.*, like the other packages?

File: proto-lens-discrimination/src/Data/ProtoLens/Discrimination.hs [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Data.ProtoLens.Discrimination
+    ( discDescriptor
+    , discField
+    , discText
+    , discByteString
+
```
Comment (judah) [inline proto-lens-discrimination/src/Data/ProtoLens/Discrimination.hs]: Remove blank line.  Or alternately, add a Haddock header to distinguish; but since this module is so small I don't know if it's needed.

File: proto-lens-discrimination/src/Data/ProtoLens/Group.hs [hunk 9]
```diff
@@ -0,0 +1,108 @@
+-- | This module implements discriminators of protobufs generically.
+
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Data.ProtoLens.Group
+    -- * Comparing Protobuf Messages
```
Comment (judah) [inline proto-lens-discrimination/src/Data/ProtoLens/Group.hs]: I think this header won't appear in the haddocks unless it's within the (..) of the export list.

File: proto-lens-discrimination/src/Data/ProtoLens/Group.hs [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    ( eqMessage
+
+    -- * 'Group's for Messages
+    , groupingMessage
+    , groupingDescriptor
+    , groupingField
+
```
Comment (judah) [inline proto-lens-discrimination/src/Data/ProtoLens/Group.hs]: Remove space or add a header?

File: proto-lens-discrimination/src/Data/ProtoLens/Sort.hs [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+{-# LANGUAGE ScopedTypeVariables #-}
+module Data.ProtoLens.Sort
+    -- * Comparing Protobuf Messages
+    ( compareMessage
+
+    -- * 'Sort's for Messages
+    , sortingMessage
+    , sortingDescriptor
+    , sortingField
+
```
Comment (judah) [inline proto-lens-discrimination/src/Data/ProtoLens/Sort.hs]: haddock comment?

File: proto-lens-discrimination/src/Data/ProtoLens/Sort.hs [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    )
+import Data.ProtoLens.Discrimination
+    ( discDescriptor
+    , discField
+    , discText
+    , discByteString
+    )
+import Data.ProtoLens.Group (groupingMessage)
+
+-- Used to adapt 'sortingCompare' to protobuf types without adding an orphan
```
Comment (judah) [inline proto-lens-discrimination/src/Data/ProtoLens/Sort.hs]: Make this a Haddock comment ("-- |")

File: proto-lens-tests/proto-lens-tests.cabal (hunk 1 above)
Comment (awpr) [inline proto-lens-tests/proto-lens-tests.cabal]: Done re: both.

File: discrimination-ieee754/test/test.hs (hunk 3 above)
Comment (awpr) [inline discrimination-ieee754/test/test.hs]: It turns out Ord is somewhat nonsensical for Float and Double because the boolean comparisons obey the requirement that NaN is incomparable to anything, and 'compare' is implemented in terms of them.  So `compare NaN x` is `GT`, but `compare x NaN` is also `GT`.  So I short-circuited out the "agree with Ord" tests for NaN arguments, added a bunch of special cases, and added tests that this matches the IEEE754 total ordering to cover the NaN cases.

File: discrimination-ieee754/src/Data/Discrimination/IEEE754.hs (hunk 4 above)
Comment (awpr) [inline discrimination-ieee754/src/Data/Discrimination/IEEE754.hs]: Resolved the question with judicious use of QuickCheck -- yes, this is IEEE754 total ordering.

File: proto-lens-discrimination/proto-lens-discrimination.cabal (hunk 6 above)
Comment (awpr) [inline proto-lens-discrimination/proto-lens-discrimination.cabal]: I'm not sure what you mean by the second part?

File: proto-lens-discrimination/src/Data/ProtoLens/Sort.hs (hunk 12 above)
Comment (awpr) [inline proto-lens-discrimination/src/Data/ProtoLens/Sort.hs]: It's not exported, so it'd be meaningless for it to be a Haddock comment.

**Comment by judah** [comment]
Closing since this PR hasn't been updated in over a year.
